RFD-0022: Key Agent Forwarding

tcsc opened the pull request:
# Overview

Proposes changing the default behaviour of `tsh ssh -A` to forward the System Key Agent to the remote machine, rather than the ephemeral Key Agent that runs inside the `tsh` process. This change would bring the behaviour of `tsh ssh` into line with the OpenSSH `ssh` client.

This builds on a proposed solution to #1517 in order to provide an escape hatch that allows legacy behaviour if necessary.

# Security Concerns

Forwarding a Key Agent is inherently a security risk as it allows anyone with sufficient
privileges on the remote to perform operations with the User's keys on the local machine.

The local keys are vulnerable for the duration of any ssh connection (with a forwarded Key Agent)
to a remote machine.

This change will also potentially expose _more_ keys to danger than the existing behaviour. The
existing `tsh ssh` Key Agent forwarding system exposes only the keys in the `~/.tsh` directory.
Allowing the user to forward their own Key Agent changes that risk to _all_ keys managed by that
agent by that. Depending on how the user invoked `tsh login`, this may also include the teleport
keys as well.

My (admittedly naive) opinion is that given forwarding the User Key Agent

 1. exposes nothing worse than OpenSSH already allows,
 2. allows the user some extra utility they would miss from OpenSSH,
 3. is a time-limited vulnerability, in that the attacker can only use the keys as long as the
    SSH connection lives, and
 4. is opt-in behaviour

...it's _probably_ OK, but I welcome points to the contrary.

tcsc committed:
RFD-0021: Key Agent Forwarding

tcsc committed:
Typo fix

tcsc committed:
Applies to any Key Agent, not just the User

xacrimon commented on rfd/0021-ssh-agent-forwarding.md:
```diff
@@ -0,0 +1,189 @@
+
+### Background: What is Key Agent Forwarding?
+
+Am SSH Key Agent is a service that brokers access to keys, used by an SSH client
```
`An`

tcsc replied:
Fixed, thanks.

xacrimon commented on rfd/0021-ssh-agent-forwarding.md:
```diff
@@ -0,0 +1,189 @@
+values be allowed for the `ForwardAgent`option:
+
+
+| Value          | Interpretation                                          |
```
Would it make sense to have a `both` option?

tcsc replied:
Yes, it probably would. 

I imagine we would have to treat it like we're forwarding the Teleport Key Agent, and have the User Key Agent be aggregated underneath that somehow. There _should_ be some way for us to delegate queries from one to the other - I'll take a look.

That said, you could simulate a `both` behaviour by adding the keys provided by `ssh login` to the user's key agent (e.g. `tsh login --add-keys-to-agent yes`) and then forwarding that with `tsh ssh -A`.

tcsc committed:
Typo fix

Also adds PlantUML diagram source as an appendix, to make it easier to
edit the diagram later.

tcsc committed:
Bump RFD number to avoid clash

tcsc committed:
Clarification & Trim

russjones commented on rfd/0022-ssh-agent-forwarding.md:
```diff
@@ -0,0 +1,224 @@
+machine to authenticate against a "Third Machine" from inside their session on the Remote
+Machine.
+
+When running `tsh ssh`, a user may have _two_ independent key agents running: their own
+(which we will call the _User Key Agent_), and an ephemeral one inside the `tsh` process
+itself (the _Teleport Key Agent_). The Teleport Key Agent is populated with the contents
+of the user's `.tsh` directory.
```
Can you check what we call this in code, I think we refer to the agent at `SSH_AUTH_SOCK` as the system agent.

tcsc replied:
Most commonly referred to as the _System Key Agent_. Have updated RFD to match.

russjones commented on rfd/0022-ssh-agent-forwarding.md:
```diff
@@ -0,0 +1,224 @@
+
+## Details
+
+### Proposed Solution
```
Can you add a section to make it clear that this does not affect auth (which will still use keys on disk), these changes are specific to agent forwarding.

tcsc replied:
Done

russjones commented on rfd/0022-ssh-agent-forwarding.md:
```diff
@@ -0,0 +1,224 @@
+`tsh ssh` will NOT forward an Key Agent to the remote machine, consistent with the behaviour of
+OpenSSH.
+
+### Security Concerns
```
I think this section can be removed.

tcsc replied:
OK, happy to remove. I think we should document the fact that we thought about the security concerns _somewhere_, though. Perhaps in the PR comment chain?

tcsc replied:
Moved security discussion to PR description

russjones commented on rfd/0022-ssh-agent-forwarding.md:
````diff
@@ -0,0 +1,224 @@
+
+```bash
+$ tsh ssh -o "ForwardAgent local" root@example.com
+```
````
Create a subsection that highlights where we break backward compatibility.

tcsc replied:
Done

russjones commented:
@tcsc Can you also add a PR description for why you are making these changes, add the original ticket and more description.

tcsc committed:
Addressing review feedback

tcsc committed:
Removed Security Concerns section

tcsc committed:
User Key Agent -> System key Agent

filipenf commented on rfd/0022-ssh-agent-forwarding.md:
```diff
@@ -0,0 +1,205 @@
+|----------------|----------------------------------------------------------|
+| `no` (default) | `tsh ssh` will not forward _any_ Key Agent               |
+| `yes`          | `tsh ssh` will forward the System Key Agent (if present) |
+| `local`        | `tsh ssh` will forward the Teleport Key Agent            |
```
I know I'm a bit too late, but I was wondering if wouldn't make sense to keep the openssh (yes/no) and use an external configuration (like an environment variable or a config file) to control weather or not to export the local agent. 

For the user, `-o ForwardAgent local` is way more verbose, introduce the precedence issue you mentioned and seem a bit counter-intuitive.
